### ISSUE TITLE
Why did you modify `greedy_search` ?

### ISSUE THREAD

**Issue by thangld201** [HojunCho/StyleBART#2]
Hi @HojunCho, In `custom/generation_utils.py`, I see that greedy search is modified (Line 103-130):
```
            ############## Only Change from this ########################
            # pre-process distribution
            next_tokens_scores = logits_processor(input_ids, next_token_logits)

            if "tokens_scores" not in model_kwargs:
                model_kwargs['tokens_scores'] = next_tokens_scores.unsqueeze(1)
            else:
                model_kwargs['tokens_scores'] = torch.cat([model_kwargs['tokens_scores'], next_tokens_scores.unsqueeze(1)], dim=1)

            # Store scores, attentions and hidden_states when required
            if return_dict_in_generate:
                if output_scores:
                    scores += (next_tokens_scores,)
                if output_attentions:
                    decoder_attentions += (
                        (outputs.decoder_attentions,) if self.config.is_encoder_decoder else (outputs.attentions,)
                    )
                    if self.config.is_encoder_decoder:
                        cross_attentions += (outputs.cross_attentions,)

                if output_hidden_states:
                    decoder_hidden_states += (
                        (outputs.decoder_hidden_states,)
                        if self.config.is_encoder_decoder
                        else (outputs.hidden_states,)
                    )

            ############## Only Change to this ########################
```

The original Transformers implementation is
```
            # pre-process distribution
            next_tokens_scores = logits_processor(input_ids, next_token_logits)

            # Store scores, attentions and hidden_states when required
            if return_dict_in_generate:
                if output_scores:
                    scores += (next_tokens_scores,)
                if output_attentions:
                    decoder_attentions += (
                        (outputs.decoder_attentions,) if self.config.is_encoder_decoder else (outputs.attentions,)
                    )
                    if self.config.is_encoder_decoder:
                        cross_attentions += (outputs.cross_attentions,)

                if output_hidden_states:
                    decoder_hidden_states += (
                        (outputs.decoder_hidden_states,)
                        if self.config.is_encoder_decoder
                        else (outputs.hidden_states,)
                    )
```

So the only part that is new is
```
            if "tokens_scores" not in model_kwargs:
                model_kwargs['tokens_scores'] = next_tokens_scores.unsqueeze(1)
            else:
                model_kwargs['tokens_scores'] = torch.cat([model_kwargs['tokens_scores'], next_tokens_scores.unsqueeze(1)], dim=1)
```

However, I do not see `'tokens_scores'` used anywhere else in the project. What bug did you encounter that you had to change the codes here ?

**Comment by HojunCho** [comment]
Before answering the question, you first need to notice that the code is based on the transformers v4.19.4 which was the latest version at that time.
https://github.com/huggingface/transformers/blob/v4.19.4/src/transformers/generation_utils.py#L1683

```python
            next_token_logits = outputs.logits[:, -1, :]

            # Store scores, attentions and hidden_states when required
            if return_dict_in_generate:
                if output_scores:
                    scores += (next_token_logits,)
                if output_attentions:
                    decoder_attentions += (
                        (outputs.decoder_attentions,) if self.config.is_encoder_decoder else (outputs.attentions,)
                    )
                    if self.config.is_encoder_decoder:
                        cross_attentions += (outputs.cross_attentions,)

                if output_hidden_states:
                    decoder_hidden_states += (
                        (outputs.decoder_hidden_states,)
                        if self.config.is_encoder_decoder
                        else (outputs.hidden_states,)
                    )

            # pre-process distribution
            next_tokens_scores = logits_processor(input_ids, next_token_logits)
```
Our model requires the probability scores after the logit processing, but on that version, the output scores were not from logits_processor but just logits.
So, I changed the codes slightly.

**Comment by thangld201** [comment]
@HojunCho Thank you for the clarification! So if I get it correctly, then this modification is not needed with the newer Transformers version I'm using (4.37.2). Plus, the following lines can be commented out as they do nothing in either versions (`tokens_scores` is not accessed elsewhere).

```
            if "tokens_scores" not in model_kwargs:
                model_kwargs['tokens_scores'] = next_tokens_scores.unsqueeze(1)
            else:
                model_kwargs['tokens_scores'] = torch.cat([model_kwargs['tokens_scores'], next_tokens_scores.unsqueeze(1)], dim=1)
```

**Comment by HojunCho** [comment]
Yes, I guess. But I couldn't remember why I implemented tokens_scores separately along with output_scores. 😕 
Maybe it is just a legacy for much older transformers version.
